### PR TITLE
[LLVM][CodeGen][AArch64] Remove bogus lowering of sve_while{gt/ge/hi/hs} intrinsics.

### PR DESCRIPTION
**Pull request by paulwalker-arm**
When fed constant operands we try to lower WHILE intrinsics to PTRUE using a fixed length pattern.  This is not valid for the decrementing variants of WHILE because they construct their result predicate vector by traversing from high->low lanes whereas the incrementing variants and PTRUE traverse from low->high lanes.

Whilst we can still utilise PTRUE by reversing its result I figure replacing a single WHILE with multiple SVE instructions is likely counterproductive.

### CODE REVIEW

**Review by dtemirbulatov — APPROVED** [review]
LGTM, With unused tests now could be removed.

**Review by dtemirbulatov — CHANGES_REQUESTED** [review]
Sorry, While revisiting this review once again examining the change, I am a bit confused now. If we look at WHILE{GT/GE/HI,HS} instructions, according to the spec it says:
...
for e = elements-1 downto 0
...
Looks like, it should not be a problem for traversing from high->low lanes?

**Comment by paulwalker-arm** [comment]
> Sorry, While revisiting this review once again examining the change, I am a bit confused now. If we look at WHILE{GT/GE/HI,HS} instructions, according to the spec it says: ... for e = elements-1 downto 0 ... Looks like, it should not be a problem for traversing from high->low lanes?

The fact the WHILE{GT/GE/HI,HS} instructions traverse from high->low lanes is what makes the replacement with PTRUE incorrect.  For example:
```
whilege(2,0) -> 0 0 0 0 0 1 1 1
```
whereas:
```
ptrue vl3    -> 1 1 1 0 0 0 0 0
```

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -5036,20 +5036,17 @@ static inline SDValue getPTrue(SelectionDAG &DAG, SDLoc DL, EVT VT,
 }
 
 static SDValue optimizeWhile(SDValue Op, SelectionDAG &DAG, bool IsSigned,
```
Comment (sdesmalen-arm) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:5038]: minor nit: is it worth renaming this to `optimizeIncrementingWhile` to distinguish it further?

**Review by sdesmalen-arm — APPROVED** [review]

**Review by dtemirbulatov — APPROVED** [review]

**Pull request merged**